### PR TITLE
Implement `vec_interleave()`

### LINKED ISSUE #1396: Add `vec_interleave()`
In `tidyr::pivot_longer()` it is at one point necessary to "interleave" a vector resp. fill its value into a matrix by row and then convert it back to a vector

```r
idx <- matrix(seq_len(nrow(data) * length(val_cols)), ncol = nrow(data), byrow = TRUE)
vec_slice(out, as.integer(idx))
```

or with a more concrete example

``` r
idx <- matrix(1:10, ncol = 2, byrow = TRUE)
vctrs::vec_slice(1:10, as.integer(idx))
#>  [1]  1  3  5  7  9  2  4  6  8 10
```

For big data this is very inefficient:

* surprisingly `matrix(x, byrow = TRUE)` is quite a bit slower than `t(matrix(x))`
* it is quite memory and time intensive just to create the index.

In [this discussion](https://github.com/tidyverse/tidyr/pull/1132#discussion_r653864349) Hadley brought up that it might make sense to move this to `vctrs` (even though it might be rare).

> Do you think it's worth pulling this out into an `interleave` function? I'm not sure this comes up very often, but we could also move into vctrs and consider our own C implementation (which I think should be fairly straight forward)

What do you think?

PS: I already asked Hadley about it but maybe you are rather the experts here. Do you have any idea if there would be any performance benefit if instead of using `vec_slice()` to reorder a vector like in this case there were a function `vec_reorder()` that only reorders the elements of it? I could imagine `vec_reorder()` to be more efficient but I don't know much about C code and the vctrs internals.

### PR DESCRIPTION
**Pull request by DavisVaughan**
Closes #1396 
Closes tidyverse/funs#10

This is essentially just a `vec_unchop()` with a fast way of generating the `indices` to unchop with.

@lionel- do you think this is worth including in vctrs? It seems reasonable to me, but is a little bit borderline because in some ways it is just fast syntactic sugar on top of `vec_unchop()`.

It has two main benefits over the previous approaches (like the one in https://github.com/tidyverse/funs/issues/10#issuecomment-538505991): less memory usage and faster index generation.

### Less memory usage

The previous approach would `vec_c()` all of the inputs in sequential order, and then reslice that result using a vector of locations that puts the elements in interleaved order. This essentially generates the full output twice, which can be expensive with data frames with many columns.

``` r
library(vctrs)
library(rlang)
library(nycflights13)

vec_interleave_new <- vctrs:::vec_interleave

vec_interleave_old <- function(..., .ptype = NULL) {
  args <- list2(...)
  n_args <- vec_size(args)
  
  if (n_args == 0L) {
    return(NULL)
  }
  
  args <- vec_recycle_common(!!! args)
  out <- vec_c(!!! args, .ptype = .ptype)
  
  pos <- vec_seq_along(out)
  pos <- as.integer(matrix(pos, nrow = n_args, byrow = TRUE))
  
  vec_slice(out, pos)
}

# 10 mil rows, 19 cols
flights <- dplyr::sample_n(flights, 1e7, replace = TRUE)

bench::mark(
  old = vec_interleave_old(flights, flights, flights),
  new = vec_interleave_new(flights, flights, flights),
  iterations = 10
)
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 old           3.42s    4.25s     0.230    6.93GB    0.321
#> 2 new           1.53s    2.13s     0.463    3.46GB    0.278
```

### Faster index generation

We use a C level `vec_interleave_indices()` to generate a list of indices. The alternative is to generate the indices at the R level, which would make the function pure R, but I feel like doing this in C is probably worth it?

``` r
library(vctrs)
library(rlang)
library(purrr)

vec_interleave_indices_c <- vctrs:::vec_interleave_indices

vec_interleave_indices_r <- function(n, size) {
  starts <- seq_len(n)
  step <- seq2(0L, size - 1L)
  map(starts, function(start) start + n * step)
}

# n = number of inputs
# size = common size
bench::mark(
  c = vec_interleave_indices_c(n = 2L, size = 10000L),
  r = vec_interleave_indices_r(n = 2L, size = 10000L)
)
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 c            8.33µs   13.9µs    46649.    78.2KB     65.4
#> 2 r          115.33µs  133.8µs     6978.   138.6KB     12.5

bench::mark(
  c = vec_interleave_indices_c(n = 50L, size = 1000000L),
  r = vec_interleave_indices_r(n = 50L, size = 1000000L)
)
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 c            33.7ms   50.5ms     19.8      191MB    79.2 
#> 2 r             270ms    270ms      3.70     195MB     3.70
```

CC @mgirlich

### CODE REVIEW

**Comment by lionel-** [comment]
This seems reasonable to me.

**Comment by DavisVaughan** [comment]
Checks will fail because it is hard to stay ahead of snapshot check changes from rlang

**Review by lionel- — APPROVED** [review]
Looks good!

But how do we undo the interleaving? Should we export the indices helper?

```r
xs <- list(letters[1:3], LETTERS[1:3])
locs <- vec_interleave_indices(2, 3)

locs
#> [[1]]
#> [1] 1 3 5
#>
#> [[2]]
#> [1] 2 4 6

vec_unchop(xs, locs)
#> [1] "a" "A" "b" "B" "c" "C"

vec_chop(vec_unchop(xs, locs), locs)
#> [[1]]
#> [1] "a" "b" "c"
#>
#> [[2]]
#> [1] "A" "B" "C"
```

File: R/interleave.R [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  vec_unchop(
+    x = args,
+    indices = indices,
+    ptype = .ptype,
+    name_spec = .name_spec,
+    name_repair = .name_repair
+  )
+}
+
+vec_interleave_indices <- function(n, size) {
```
[the remark this answers is not in the document]
Comment (lionel-) [reply R/interleave.R:64]: It's possible this function could be made more general. It gives the (chopped) indices to switch from column major order to row major order, or equivalently to transpose 2-dimensional data. Not sure how to deal with higher dimensions though.

Maybe `vec_transpose_locs()` if we keep it to 2 dimensions?

[the remark this answers is not in the document]
Comment (lionel-) [reply R/interleave.R:64]: There's a relationship between `vec_interleave()` and `vec_zip()` / `vec_transpose()` so we might want to consider these together.

[the remark this answers is not in the document]
Comment (DavisVaughan) [reply R/interleave.R:64]: `vec_zip()` / `vec_transpose()` would be tough functions to do efficiently without a specialized C function that can do all of the slicing and assigning in one step.

Based on the definitions in https://observablehq.com/@d3/d3-transpose I came up with the implementations below, making `vec_interleave()` just an unchop of the transposed result - but interleaving directly is much easier to do.

``` r
library(vctrs)
library(rlang)
library(purrr)

vec_transpose <- function(x, 
                          ptype = NULL,
                          name_spec = NULL,
                          name_repair = c("minimal", "unique", "check_unique", "universal")) {
  if (!vec_is_list(x)) {
    abort("`x` must be a list to transpose.")
  }
  
  missing <- vec_equal_na(x)
  if (any(missing)) {
    x <- vec_slice(x, !missing)
  }
  
  x <- vec_recycle_common(!!!x)
  ptype <- vec_ptype_common(!!!x, .ptype = ptype)
  
  x <- map(x, vec_chop)
  
  pmap(
    .l = x, 
    .f = vec_c, 
    .ptype = ptype, 
    .name_spec = name_spec, 
    .name_repair = name_repair
  )
}

vec_zip <- function(...,
                    .ptype = NULL,
                    .name_spec = NULL,
                    .name_repair = c("minimal", "unique", "check_unique", "universal")) {
  vec_transpose(list2(...), ptype = .ptype, name_spec = .name_spec, name_repair = .name_repair)
}

vec_zip(x = 1:3, y = 4:6, z = 7:9)
#> [[1]]
#> x y z 
#> 1 4 7 
#> 
#> [[2]]
#> x y z 
#> 2 5 8 
#> 
#> [[3]]
#> x y z 
#> 3 6 9

vec_zip(list(1, 2, 3), list(4, 5, 6))
#> [[1]]
#> [[1]][[1]]
#> [1] 1
#> 
#> [[1]][[2]]
#> [1] 4
#> 
#> 
#> [[2]]
#> [[2]][[1]]
#> [1] 2
#> 
#> [[2]][[2]]
#> [1] 5
#> 
#> 
#> [[3]]
#> [[3]][[1]]
#> [1] 3
#> 
#> [[3]][[2]]
#> [1] 6
```

[the remark this answers is not in the document]
Comment (DavisVaughan) [reply R/interleave.R:64]: I'm not sure that it is worth going past 2 dimensions here. I could see how something like `aperm()` that just returned the indices could be useful in an array specific package though.

**Comment by DavisVaughan** [comment]
> But how do we undo the interleaving? Should we export the indices helper?

I briefly considered this but decided against it because I have never seen any need to undo it. In the few times that we've done any kind of interleaving, it is always one directional.

Do you mind if we just export `vec_interleave()` for now, and revisit if the need arises?

**Comment by lionel-** [comment]
> Do you mind if we just export vec_interleave() for now, and revisit if the need arises?

That sounds good.

**Pull request merged**